### PR TITLE
feat: added info page

### LINKED ISSUE #850: frontend APISIX instance status page
# Feature request

## Please describe your feature
We have already some discussions about this topic: https://lists.apache.org/thread.html/re3097f7e88d636e8137231c36462c32773b2b4817d91f7997e91358f%40%3Cdev.apisix.apache.org%3E.

## Additional context

This thread is used to track the development progress of Dashboard frontend part.

The related issues:

* https://github.com/apache/apisix/issues/2821
* apache/apisix-dashboard#849

### PR DESCRIPTION
**Pull request by guoqqqi**
### TODO

- [x] Add Docs https://github.com/apache/apisix/blob/master/doc/plugins/server-info.md

resolve #850

I have made a static page at present. Is the page style like this? I have some questions and need help:
1 When I add an icon to the Info page in the antd <IconFont> component, I don't seem to have permission to add icons and view existing icons. So I used an antd icon, but it didn't show up. At the top of the menu bar!
2 The page style is currently written inline. If there is no problem with the style, I will write it in styles.less.
3 The rest is the <select> dynamic display table, which is being done.

### CODE REVIEW

**Comment by juzhiyuan** [comment]
This PR still looks like a draft, let's convert it as draft.

**Comment by juzhiyuan** [comment]
image

for this check, you need to add the Apache Software Foundation's License header at the top of your changed files, please refer to other pages.

File: web/src/pages/Info/List.tsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import React from 'react';
+import { PageContainer } from '@ant-design/pro-layout';
+import { useIntl } from 'umi';
+
+const Info: React.FC = () => {
+  const { formatMessage } = useIntl();
+
+  return (
+    <PageContainer title={formatMessage({ id: 'page.info.pageContainer.title' })}>
+      <select className="selector" id="selector" style={{ backgroundColor: "#fff", width: "100%", height: "24px", textAlign: "right", padding: "16px 24px", marginBottom: "15px" }}>
```
Comment (juzhiyuan) [inline web/src/pages/Info/List.tsx:10]: 1. Use antd's Selector component;

File: web/src/pages/Info/List.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        <option value="categoryA">CategoryA</option>
+        <option value="categoryB">CategoryB</option>
+        <option value="categoryC">CategoryC</option>
+      </select>
+      <div className="wrap" style={{ width: "100%", height: "740px", padding: "16px 24px", margin: "0 auto", backgroundColor: "#fff" }}>
```
Comment (juzhiyuan) [inline web/src/pages/Info/List.tsx:15]: Using div and other HTML tags look good to me, but you should move those inline style to a standalone file.

File: web/src/pages/Info/List.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        </table>
+        <table style={{ color: "rgba(0, 0, 0, 0.45)", width: "100%", borderBottom: "1px solid grey", borderCollapse: "collapse", }}>
+          <thead>
+            <tr>
+              <th style={{ padding: "12px 8px" }}>categoryA</th>
+              <th></th>
+            </tr>
+          </thead>
+          <tbody>
+            <tr style={{ backgroundColor: "#f3f4f5" }}>
```
Comment (juzhiyuan) [inline web/src/pages/Info/List.tsx:76]: Why not use an array loop?

File: web/src/pages/Info/locales/en-US.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export default {
+    'page.info.pageContainer.title': 'About Apache APISIX',
```
Comment (juzhiyuan) [inline web/src/pages/Info/locales/en-US.ts:2]: 2 spaces

File: web/src/pages/Info/styles.less [hunk 5]
```diff
@@ -0,0 +1,6 @@
+@import '~antd/es/style/themes/default.less';
+
+//未引入，将内联css放这里，待执行。
```
Comment (juzhiyuan) [inline web/src/pages/Info/styles.less:3]: Only English is allowed for comments.

**Review by juzhiyuan — CHANGES_REQUESTED** [review]

**Comment by juzhiyuan** [comment]
Any update?

**Comment by guoqqqi** [comment]
> Any update?

Wait a minute, I will update the progress.

**Comment by guoqqqi** [comment]
At present, an onChange function needs to be added to make the Select component and the table linkage.

File: web/src/pages/Info/List.tsx [hunk 6]
```diff
@@ -1,96 +1,92 @@
...
-            </tr>
-            <tr>
-              <td style={{ padding: "12px 8px" }}>key</td>
-              <td style={{ textAlign: "right", padding: "12px 8px" }}>value</td>
-            </tr>
-          </tbody>
-        </table>
+      <Select
+        showSearch
+        style={{ backgroundColor: "#fff", width: "100%", textAlign: "right", padding: "16px 24px", marginBottom: "15px" }}
```
Comment (juzhiyuan) [inline web/src/pages/Info/List.tsx:60]: Inline styles should be moved to a `.less`file

File: web/src/global.less [hunk 7]
```diff
@@ -45,6 +45,27 @@ ol {
   list-style: none;
 }
 
+.InfoWrap {
```
Comment (juzhiyuan) [inline web/src/global.less:48]: No, this is a global file, not for Info page only.

**Review by juzhiyuan — CHANGES_REQUESTED** [review]
Please paste the snapshot here.

File: web/config/routes.ts [hunk 8]
```diff
@@ -23,6 +23,10 @@ const routes = [
     path: '/metrics',
     component: './Metrics',
   },
+  {
+    path: '/info/list',
```
Comment (juzhiyuan) [inline web/config/routes.ts:27]: `/info` is enough

File: web/src/helpers.tsx [hunk 9]
```diff
@@ -22,9 +22,15 @@ import moment from 'moment';
 
 import { codeMessage } from './constants';
 import IconFont from './iconfont';
+import InfoCircleOutlined from '@ant-design/icons';
```
Comment (juzhiyuan) [inline web/src/helpers.tsx:25]: move this line to the top

File: web/src/pages/Info/List.tsx [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+ */
+import React from 'react';
+import { Select } from 'antd';
+import { PageContainer } from '@ant-design/pro-layout';
+import { useIntl } from 'umi';
+
+const Info: React.FC = () => {
+  const { formatMessage } = useIntl();
+  const { Option } = Select;
+  const mydata = [
```
Comment (juzhiyuan) [inline web/src/pages/Info/List.tsx:25]: Why not use useState? then you could update it by using setXXX

**Review by juzhiyuan — CHANGES_REQUESTED** [review]

**Comment by guoqqqi** [comment]
WeChat82997997eab173ce6f069a584829fb60
WeChatbd644c596d6bd3c38275ea2084d094d5
1. After the change, the icon of the information page in the menu bar is still not displayed.
2. After debugging, click <Select> to search, and <table> is displayed, whether the style needs further optimization.

**Comment by juzhiyuan** [comment]
This style looks good to me, i will check why that info icon is missing.

Any suggestion for this style? 

cc @moonming @liuxiran @tokers @starsz

**Comment by juzhiyuan** [comment]
cc @guoqqqi This PR's checks have failed, just click the details link to see what's happened.

image

File: web/src/helpers.tsx [hunk 11]
```diff
@@ -20,11 +20,17 @@ import { MenuDataItem } from '@ant-design/pro-layout';
 import { history } from 'umi';
 import moment from 'moment';
 
+import InfoCircleOutlined from '@ant-design/icons';
```
Comment (juzhiyuan) [inline web/src/helpers.tsx:23]: 1. image

2. You may use the wrong import method

```ts
import {InfoCircleOutlined} from '@ant-design/icons';
```

File: web/src/pages/Info/List.tsx [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    value: 'value2',
+  },
+  {
+    key: 'key3',
+    value: 'value3',
+  },
+];
+
+const Info: React.FC = () => {
+  const [detail, setDetail] = useState([] as any);
```
Comment (juzhiyuan) [inline web/src/pages/Info/List.tsx:71]: ugly typing!

File: web/src/pages/Info/List.tsx [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+  const [nodeList, setNodeList] = useState([] as any);
+  const { formatMessage } = useIntl();
+  const { Option } = Select;
+
+  useEffect(() => {
+    setNodeList(nodeListData);
+  });
```
Comment (juzhiyuan) [inline web/src/pages/Info/List.tsx:78]: what's the difference when using `[]` and without `[]`?

File: web/src/pages/Info/List.tsx [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+
+  return (
+    <PageContainer title={formatMessage({ id: 'page.info.pageContainer.title' })}>
+      <div className={styles.select}>
+        <Select
+          showSearch
+          style={{ width: '130px' }}
+          placeholder="Please select"
```
Comment (juzhiyuan) [inline web/src/pages/Info/List.tsx:86]: select what?

File: web/src/pages/Info/List.tsx [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+          onChange={() => {
+            setDetail(detailData);
```
Comment (juzhiyuan) [inline web/src/pages/Info/List.tsx:88]: call api then set data when api is ready

File: web/src/pages/Info/List.tsx [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+          }}
+        >
+          {nodeList.map((item: any) => (
```
Comment (juzhiyuan) [inline web/src/pages/Info/List.tsx:91]: why any?

File: web/src/pages/Info/List.tsx [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+            <Option value={item.key} lable={item.key}>
```
Comment (juzhiyuan) [inline web/src/pages/Info/List.tsx:92]: use key for loop

<Option key={xxx} />

File: web/src/pages/Info/List.tsx [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+              {item.key}
+            </Option>
+          ))}
+          ;
```
Comment (juzhiyuan) [inline web/src/pages/Info/List.tsx:96]: ?

File: web/src/pages/Info/List.tsx [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+                <th>&nbsp;</th>
+              </thead>
+              <tbody>
+                <tr>
+                  <td>{item.name}</td>
+                  <td>{item.value}</td>
+                </tr>
+                <tr>
+                  <td>{item.name2}</td>
+                  <td>{item.value2}</td>
```
Comment (juzhiyuan) [inline web/src/pages/Info/List.tsx:114]: 2?

File: web/src/pages/Info/List.tsx (hunk 12 above)
Comment (juzhiyuan) [inline web/src/pages/Info/List.tsx:71]: bad variable name, use `data` here.

**Review by juzhiyuan — CHANGES_REQUESTED** [review]

**Comment by guoqqqi** [comment]
WeChate6a77b6a1a56e4931dadb01f48b59a2a
WeChate114c86b55ebdea37f60761dac054250
1. Update the selected node and the corresponding information appears. Wait for the api interface, and then update asynchronously to obtain data.
2. At present, when the title of the table is first loaded, the data page will crash if it is not obtained. I will comment it out first and I will try to solve it tomorrow.

**Comment by juzhiyuan** [comment]
emm selector is used to select nodes, not category.

**Comment by juzhiyuan** [comment]
## APIs

### Fetch Node List
TODO

### Fetch Node Detail
```json
{
    "id": "ddea4053-4ffd-4f09-b176-b8c5e919916b",
    "hostname": "localhost.local",
    "version": "2.0",
    "etcd_version": "3.5.0",
    "uptime": 3600,
    "last_report_at": 1606121591,
    "boot_time": 1606121501
}
```

**Comment by starsz** [comment]
### Get server_info
Method: Get
URI: /apisix/server_info/{id}
Response:
```
{
    "code":0,
    "message":"",
    "data":{
        "id":1,
        "last_report_time":1607400888,
        "up_time":100,
        "boot_time":1607400745,
        "etcd_version":"3.5.0",
        "hostname":"gentoo",
        "version":"2.0"
    },
    "request_id":"afd668da-5245-40c4-a674-011014a22753"
}
```

### list server_info
Method: Get
URI: /apisix/server_info?hostname={hostname}&page_size={page_size}&page={page}
Response:
```
{
    "code":0,
    "message":"",
    "data":{
        "rows":[
            {
                "id":1,
                "last_report_time":1607400888,
                "up_time":100,
                "boot_time":1607400745,
                "etcd_version":"3.5.0",
                "hostname":"gentoo",
                "version":"2.0"
            }
        ],
        "total_size":2
    },
    "request_id":"66f1a516-3ed9-411a-a8dc-9d5cc130fedc"
}
```

|field|value|note|
|---|---|---|
|hostname|default ""|If the hostname is not "", the API only returns the server_info whose `hostname` field contains hostname.
|page_size|default 0|If the `page_size` is larger than zero. The API only returns the `page_size` number value of data per request.
|page|default 0|If the `page` is lager than zero.The api only returns the [(page-1)\*page_size, page\*page_size).Otherwise it will return whole server_infos.
----

BTW, I add `hostname` for filter response if there are lots of apisix instances

**Comment by juzhiyuan** [comment]
https://github.com/apache/apisix-dashboard/pull/1076

We use `/apisix/admin` as prefix.

File: web/src/pages/Info/locales/en-US.ts [hunk 20]
```diff
@@ -15,5 +15,5 @@
  * limitations under the License.
  */
 export default {
-  'page.info.pageContainer.title': 'About Apache APISIX',
+  'page.info.pageContainer.title': 'Server info',
```
Comment (juzhiyuan) [inline web/src/pages/Info/locales/en-US.ts]: Info

File: web/src/pages/Info/service.ts [hunk 21]
```diff
@@ -19,14 +19,14 @@ import { request } from 'umi';
 // Waiting for api.
 export const fetchInfoList = () => {
   return request<Res<ResListData<NodeListData>>>(
-    '/server_info?hostname={hostname}&page_size={page_size}&page={page}',
+    `/server_info`,
```
Comment (juzhiyuan) [inline web/src/pages/Info/service.ts:22]: ''

File: web/src/pages/Info/service.ts [hunk 22]
```diff
@@ -19,14 +19,14 @@ import { request } from 'umi';
...
   ).then(({ data }) => ({
-    data: data.rows,
+    data: data.rows
   }));
 };
 
 export const fetchInfoData = () => {
```
Comment (juzhiyuan) [inline web/src/pages/Info/service.ts:26]: remove this

File: web/src/pages/Info/typing.d.ts [hunk 23]
```diff
@@ -15,16 +15,21 @@
...
+  etcd_version: string;
   hostname: string;
   version: string;
-  etcd_version: string;
-  uptime: number;
-  last_report_at: number;
-  boot_time: number;
 };
 
 type NodeListData = {
```
Comment (juzhiyuan) [inline web/src/pages/Info/typing.d.ts:27]: ?

**Review by juzhiyuan — CHANGES_REQUESTED** [review]

**Review by juzhiyuan — APPROVED** [review]
LGTM

File: web/src/pages/ServerInfo/List.tsx [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+      <div className={styles.select}>
+        <Form form={form}>
+          <Form.Item wrapperCol={{ span: 5 }} style={{ marginBottom: 0 }} name="nodeId">
+            <Select
+              placeholder={formatMessage({ id: 'page.serverinfo.select.placeholder' })}
+              onChange={(value) => {
+                setData(
+                  nodeList.find((item) => {
+                    return item.id === value;
+                  }),
```
Comment (LiteSun) [inline web/src/pages/ServerInfo/List.tsx:57]:
```suggestion
                  })
```

File: web/src/pages/ServerInfo/locales/en-US.ts [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+export default {
+  'page.serverinfo.pageContainer.title': 'Server Info',
+  'page.serverinfo.select.placeholder': 'Please select node',
+  'page.serverinfo.desc': 'The relevant plugin need to be enabled to report server info.',
+  'page.serverinfo.link': ' How to enable?',
```
Comment (LiteSun) [inline web/src/pages/ServerInfo/locales/en-US.ts:21]:
```suggestion
  'page.serverinfo.link': 'How to enable?',
```
why whitespace here?

Comment (guoqqqi) [reply web/src/pages/ServerInfo/locales/en-US.ts]: update

**Comment by juzhiyuan** [comment]
E2E failed @nic-chen

**Review by LiteSun — APPROVED** [review]
LGTM

**Pull request merged**